PS-7640 Document ProcFS plugin (8.0) (#4315)
percona/percona-server: doc/source/performance/procfs-plugin.rst

.. procfs-plugin:

======================================================================
The ProcFS plugin
======================================================================

> This feature is **tech preview** quality.

Reviewer: How about making this a `note`, so it's more prominent?

Author: Done

> Implemented in :rn:`8.0.25-15`, the ProcFS plugin provides access to the Linux performance counters by running SQL queries against a Percona Server for MySQL 8.0 or another MySQL 8.0-compatible variant. 

Reviewer: According to some recent feedback from @georgelorch , we do not guarantee plugin compatibility with other MySQL variants other than Percona Server (and PXC). So I'd suggest to remove the "and another MySQL 8.0-compatible variant" here.

Author: Done.

Author: Done.

> In some environments, you are unable to capture operating system metrics , for example, Cloud installations or MySQL-as-a-service installations. These metrics are essential for complete system performance monitoring.

Reviewer: ```suggestion
In some environments, you may be unable to capture operating system metrics , for example, Cloud installations or MySQL-as-a-service installations. These metrics are essential for complete system performance monitoring.
```

Author: Fixed.

The plugin does the following:

* Reads selected files from the ``/proc`` file system and the ``/sys`` file system.

> * Populates the file names, and any content as rows in the `INFORMATION_SCHEMA.PROCFS`.

Reviewer: ```suggestion
* Populates the file names, and their content as rows in the `INFORMATION_SCHEMA.PROCFS` view.
```

Author: Fixed.

The system variable :ref:`procfs_files_spec` provides access to the ``/proc`` and the ``/sys`` files and directories. This variable cannot be changed at run time, preventing a compromised account from giving itself greater access to those file systems. 

Install the PLUGIN
--------------------

> To install this plugin, copy the procfs.so file to the mysql plugin installation directory and execute the following command:

Reviewer: Is that the default installation method? Shouldn't we default to giving instructions on how to install the necessary package instead? Manually copying a shared object file sounds cumbersome to me.

Author: Updated.

.. sourcecode:: mysql

    INSTALL PLUGIN procfs SONAME 'procfs.so';

Access Privileges Required 
 ---------------------------

> Only users with the `ACCESS_PROCFS` dynamic privilege can access the INFORMATION_SCHEMA.PROCFS view. During the plugin startup, this dynamic privilege is registered with the server. 

Reviewer: Use RST syntax for monospaced font here:
```suggestion
Only users with the ``ACCESS_PROCFS`` dynamic privilege can access the ``INFORMATION_SCHEMA.PROCFS`` view. During the plugin startup, this dynamic privilege is registered with the server. 
```

Author: Added.

After the plugin installation, grant a user access to the INFORMATION_SCHEMA.PROCFS view by executing the following command:

.. sourcecode:: mysql

    GRANT ACCESS_PROCFS ON *.* TO 'user'@'host';

> Using the ProcFS-plugin

Reviewer: Without the dash?
```suggestion
Using the ProcFS plugin
```

Author: Fixed.

    SELECT * FROM INFORMATION_SCHEMA.PROCFS WHERE FILE = '/proc/version';

Tables
-----------

.. _PROCFS:

.. rubric:: PROCFS

The schema definition of the INFORMATION_SCHEMA.PROCFS view is:

.. sourcecode:: mysql

    CREATE TEMPORARY TABLE `PROCFS` (
    `FILE` varchar(1024) NOT NULL DEFAULT '',
    `CONTENTS` longtext NOT NULL
    ) ENGINE=InnoDB DEFAULT CHARSET=utf8;
 
Status variables provide the basic metrics:

.. list-table::
    :widths: 10 40
    :header-rows: 1

    * - Name 
      - Description
    * - procfs_access_violations
      - The number of attempted queries by users without the ACCESS_PROCFS privilege.
    * - procfs_queries
      - The number of queries made against the procfs view.
    * - procfs_files_read
      - The number of files read to provide content
    * - procfs_bytes_read
      - The number of bytes read to provide content
      

Variable
---------

.. _procfs_files_spec:

.. rubric:: `procfs_files_spec`

.. list-table::
    :widths: 20 40
    :header-rows: 1

    * - Parameter
      - Description
    * - Introduced
      - 8.0.25-14
    * - Dynamic
      - Yes
    * - Scope
      - Global
    * - Read, Write, or Read-Only
      - Read-Only
  

> The default value for `procfs_files_spec` is: /proc/cpuinfo;/proc/irq//;/proc/loadavg/proc/net/dev;/proc/net/sockstat;/proc/net/sockstat_rhe4;/proc/net/tcpstat;/proc/self/net/netstat;/proc/self/stat;/proc/self/io;/proc/self/numa_maps/proc/softirqs;/proc/spl/kstat/zfs/arcstats;/proc/stat;/proc/sys/fs/file-nr;/proc/version;/proc/vmstat

Reviewer: ```suggestion
The default value for ``procfs_files_spec`` is: /proc/cpuinfo;/proc/irq//;/proc/loadavg/proc/net/dev;/proc/net/sockstat;/proc/net/sockstat_rhe4;/proc/net/tcpstat;/proc/self/net/netstat;/proc/self/stat;/proc/self/io;/proc/self/numa_maps/proc/softirqs;/proc/spl/kstat/zfs/arcstats;/proc/stat;/proc/sys/fs/file-nr;/proc/version;/proc/vmstat
```

Author: Updated.

.. procfs-plugin:

======================================================================
Using the ProcFS plugin
======================================================================

> Implemented in |Percona Server| 8.0.25-15, the ProcFS plugin provides access to the Linux performance counters by running SQL queries against a Percona Server 8.0 (or other MySQL 8.0-compatible variant). This plugin is useful for Cloud and MySQL-as-service installations.

Reviewer 2: MySQL as a service?
```suggestion
Implemented in |Percona Server| 8.0.25-15, the ProcFS plugin provides access to the Linux performance counters by running SQL queries against  Percona Server for MySQL 8.0 (or other MySQL 8.0-compatible variant). This plugin is useful for Cloud and MySQL-as-a-service installations.
```

Author: fixed.

> Implemented in |Percona Server| 8.0.25-15, the ProcFS plugin provides access to the Linux performance counters by running SQL queries against a Percona Server for MySQL 8.0 (or other MySQL 8.0-compatible variant). This plugin is useful for Cloud and MySQL-as-a-service installations.

Reviewer 3: use :rn:`8.0.25-15` so it automatically links to release notes. This helps user to see when (date) this has been released and what other features/bugs landed at the same time.

Author: Fixed.

> Implemented in |Percona Server| 8.0.25-15, the ProcFS plugin provides access to the Linux performance counters by running SQL queries against a Percona Server for MySQL 8.0 (or other MySQL 8.0-compatible variant). This plugin is useful for Cloud and MySQL-as-a-service installations.

Reviewer 3: Can you please provide more information on this
> This plugin is useful for Cloud and MySQL-as-a-service installations

How it's useful?

Author: Added information - basically, it allows a user to access metrics when they cannot access the operating system.

> The plugin reads files from /proc and /sys file systems and returns the files name and any content in rows in the `INFORMATION_SCHEMA.PROCFS` view. The schema definition of this view is:

Reviewer 2: files names?

Author: fixed.

> 

Reviewer 3: It would be more consistent if we used same template that is used for other I_S tables, like: 
 diagnostics/misc_info_schema_tables.rst

Author: Same problem as :variable: - these tables will be reformatted so that we can remove psdom.

Author: Added link to I_S tables doc.

Reviewer 4: "in some environments" repeated twice in same sentence.

Author: Fixed.

.. sourcecode:: mysql

    CREATE TEMPORARY TABLE `PROCFS` (
    `FILE` varchar(1024) NOT NULL DEFAULT '',
    `CONTENTS` longtext NOT NULL
    ) ENGINE=InnoDB DEFAULT CHARSET=utf8;
 

> Access to /proc and /sys files and directories is enabled through the system variable `procfs_files_spec`. This variable is global and read only and is set by using either the `mysqld` command line or by editing the :file:`my.cnf`. 

Reviewer 3: did we stop using the :variable:`procfs_files_spec` syntax to automatically link to variable description.

Variable description for `procfs_files_spec` is also missing.

Author: Yes. We are dropping the :variable: syntax because documentation is dropping psdom.py.

The variable cannot be changed at run time to prevent a compromised account from giving itself greater access to the /proc or /sys file systems. 

> Any files that are not included within the `procfs_files_spec` are ignored and considered to not exist You can write file access patterns in `glob(7) style <https://man7.org/linux/man-pages/man7/glob.7.html>`__:  /sys/block/sd[a-z]/stat;/proc/version*

Reviewer 2: ```suggestion
Any files that are not included within the `procfs_files_spec` are ignored and considered to not exist. You can write file access patterns in `glob(7) style <https://man7.org/linux/man-pages/man7/glob.7.html>`__:  /sys/block/sd[a-z]/stat;/proc/version*
```

Author: fixed.

> The default value for procfs_files_spec is: /proc/cpuinfo;/proc/irq//;/proc/loadavg/proc/net/dev;/proc/net/sockstat;/proc/net/sockstat_rhe4;/proc/net/tcpstat;/proc/self/net/netstat;/proc/self/stat;/proc/self/io;/proc/self/numa_maps/proc/softirqs;/proc/spl/kstat/zfs/arcstats;/proc/stat;/proc/sys/fs/file-nr;/proc/version;/proc/vmstat

Reviewer 2: ```suggestion
The default value for `procfs_files_spec` is: /proc/cpuinfo;/proc/irq//;/proc/loadavg/proc/net/dev;/proc/net/sockstat;/proc/net/sockstat_rhe4;/proc/net/tcpstat;/proc/self/net/netstat;/proc/self/stat;/proc/self/io;/proc/self/numa_maps/proc/softirqs;/proc/spl/kstat/zfs/arcstats;/proc/stat;/proc/sys/fs/file-nr;/proc/version;/proc/vmstat
```

Author: fixed.

.. rubric:: Install plugin

> To install this plugin, copy the procfs.so file to the mysql plugin installation directory and execute 

Reviewer 3: Does this apply to packages as well? In that case where should we get the file `procfs.so` file from?

Author: This is the same type of information for each plugin.

Author: updated.

.. sourcecode:: mysql

    INSTALL PLUGIN procfs SONAME 'procfs.so';

.. rubric:: Grant Access to users

> Only users with the `ACCESS_PROCFS` dynamic privilege can access the INFORMATION_SCHEMA.PROCFS view. During the plugin startup, this dynamic privilege is registered with the server. After the plugin has been installed, you can grant a user access to the INFORMATION_SCHEMA.PROCFS view by executing the following:

Reviewer 3: there is no `INFORMATION_SCHEMA.PROCFS` in the index_info_schema_tables.rst

Author: Added to doc.
